client.go

```diff
@@ -123,10 +136,31 @@ func NewClient() *Client {
 	}
 }
 
+// DefaultRetryPolicy provides a default callback for Client.CheckRetry, which
+// will retry on connection errors and server errors.
+func DefaultRetryPolicy(resp *http.Response, err error) (bool, error) {
+	if err != nil {
+		return true, err
+	}
+	// Check the response code. We retry on 500-range responses to allow
+	// the server time to recover, as 500's are typically not permanent
+	// errors and may relate to outages on the server side. This will catch
+	// invalid response codes as well, like 0 and 999.
+	if resp.StatusCode == 0 || resp.StatusCode >= 500 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
 // Do wraps calling an HTTP method with retries.
 func (c *Client) Do(req *Request) (*http.Response, error) {
 	c.Logger.Printf("[DEBUG] %s %s", req.Method, req.URL)
 
+	if c.CheckRetry == nil {
```

Comment:
Yes, the split in functionality isn't good, and we should setup everything in one place. I like checking the fields in `Do()`, but there's no good way to check for unset values for `RetryWaitMin` and `RetryMax`which could have been set to 0 by the caller without ill effect.

So now that I think about it, since `Logger` will panic if it's not set currently, I'll move the CheckRetry into the NewClient function too.